src/core/unittest/SettingsTest.cpp

```diff
@@ -334,6 +336,277 @@ TEST(SettingsTest, StreamRecvWindowDefaultGetsOverridenByIndividualLimits)
+    ASSERT_EQ(Settings.StreamRecvWindowBidiRemoteDefault, QUIC_DEFAULT_STREAM_FC_WINDOW_SIZE);
+    ASSERT_EQ(Settings.StreamRecvWindowUnidiDefault, QUIC_DEFAULT_STREAM_FC_WINDOW_SIZE);
+    ASSERT_EQ(Settings.StreamRecvBufferDefault, QUIC_DEFAULT_STREAM_RECV_BUFFER_SIZE);
+    ASSERT_EQ(Settings.ConnFlowControlWindow, QUIC_DEFAULT_CONN_FLOW_CONTROL_WINDOW);
+    ASSERT_EQ(Settings.MaxBytesPerKey, QUIC_DEFAULT_MAX_BYTES_PER_KEY);
+    ASSERT_EQ(Settings.ServerResumptionLevel, (uint8_t)QUIC_DEFAULT_SERVER_RESUMPTION_LEVEL);
+    ASSERT_EQ(Settings.VersionNegotiationExtEnabled, QUIC_DEFAULT_VERSION_NEGOTIATION_EXT_ENABLED);
+    ASSERT_EQ(Settings.MinimumMtu, QUIC_DPLPMTUD_DEFAULT_MIN_MTU);
+    ASSERT_EQ(Settings.MaximumMtu, QUIC_DPLPMTUD_DEFAULT_MAX_MTU);
+    ASSERT_EQ(Settings.MtuDiscoveryMissingProbeCount, QUIC_DPLPMTUD_MAX_PROBES);
+    ASSERT_EQ(Settings.MtuDiscoverySearchCompleteTimeoutUs, QUIC_DPLPMTUD_RAISE_TIMER_TIMEOUT);
+    ASSERT_EQ(Settings.MaxBindingStatelessOperations, QUIC_MAX_BINDING_STATELESS_OPERATIONS);
+    ASSERT_EQ(Settings.StatelessOperationExpirationMs, QUIC_STATELESS_OPERATION_EXPIRATION_MS);
+    ASSERT_EQ(Settings.CongestionControlAlgorithm, QUIC_CONGESTION_CONTROL_ALGORITHM_DEFAULT);
+    ASSERT_EQ(Settings.DestCidUpdateIdleTimeoutMs, QUIC_DEFAULT_DEST_CID_UPDATE_IDLE_TIMEOUT_MS);
+    ASSERT_EQ(Settings.GreaseQuicBitEnabled, QUIC_DEFAULT_GREASE_QUIC_BIT_ENABLED);
+    ASSERT_EQ(Settings.EcnEnabled, QUIC_DEFAULT_ECN_ENABLED);
+    ASSERT_EQ(Settings.HyStartEnabled, QUIC_DEFAULT_HYSTART_ENABLED);
+    ASSERT_EQ(Settings.EncryptionOffloadAllowed, QUIC_DEFAULT_ENCRYPTION_OFFLOAD_ALLOWED);
+    ASSERT_EQ(Settings.ReliableResetEnabled, QUIC_DEFAULT_RELIABLE_RESET_ENABLED);
+    ASSERT_EQ(Settings.XdpEnabled, QUIC_DEFAULT_XDP_ENABLED);
+    ASSERT_EQ(Settings.QTIPEnabled, QUIC_DEFAULT_QTIP_ENABLED);
+    ASSERT_EQ(Settings.RioEnabled, QUIC_DEFAULT_RIO_ENABLED);
+    ASSERT_EQ(Settings.OneWayDelayEnabled, QUIC_DEFAULT_ONE_WAY_DELAY_ENABLED);
+    ASSERT_EQ(Settings.NetStatsEventEnabled, QUIC_DEFAULT_NET_STATS_EVENT_ENABLED);
+    ASSERT_EQ(Settings.StreamMultiReceiveEnabled, QUIC_DEFAULT_STREAM_MULTI_RECEIVE_ENABLED);
+}
+
+TEST(SettingsTest, QuicSettingsSetDefault_DoesNotOverwriteSetFields)
+{
+    QUIC_SETTINGS_INTERNAL Settings;
+    CxPlatZeroMemory(&Settings, sizeof(Settings));
+
+    // Set a few fields and mark them as set
+    Settings.IsSet.SendBufferingEnabled = 1;
+    Settings.SendBufferingEnabled = 0;
+    Settings.IsSet.PacingEnabled = 1;
+    Settings.PacingEnabled = 0;
+    QuicSettingsSetDefault(&Settings);
+
+    // These should not be overwritten
+    ASSERT_EQ(Settings.SendBufferingEnabled, 0);
+    ASSERT_EQ(Settings.PacingEnabled, 0);
+
+    // But an unset field should be set to default
+    ASSERT_EQ(Settings.MigrationEnabled, QUIC_DEFAULT_MIGRATION_ENABLED);
+}
+
+struct QuicStorageSettingScopeGuard {
+public:
+    QuicStorageSettingScopeGuard(
```

Comment:
I think there would be more value to this scope guard if it would delete recursively the entire test registry key (all values).

It is generally better to have the scope guard be returned by the call that create the resource (so they can't be out of sync), so I would suggest to:

- have a wrapper on `CxPlatStorageOpen` that returns the scope guard
- the scope guard delete all keys + closes the storage
- now, test can invoke the wrapper to open the storage, store the returned value in a local variable and not have to care about cleanup ever again.

Reply:
there's a couple nuances here:
1) the global reg state for MsQuic is a registry key that shouldn't be deleted (but the values under it can be deleted)
2) the per-app reg state is a subkey under the global key, so recursively deleted the global state would also wipe out all per-app settings (undesired)

 I designed this around the per-value scope guard because kernel mode APIs for clearing an entire reg key of all values is complex